### DISCUSSION TITLE
Reader.tile defaults to reading 3 bands instead of all bands

### DISCUSSION THREAD

**Discussion by hambsch** [cogeotiff/rio-tiler#771] [Q&A]
Hi,
while processing RGBA geotiffs, i noticed that the reader.tile does not read all bands by default, rasterio does this:

```
import rasterio
from rio_tiler.io import Reader

file_path = r"C:\RMAbuild\Projects\weiss-geoloket_api\data\route\ant-all-all-50-netto_2010.tif"

with Reader(file_path) as cog:
    img = cog.tile(2100, 1369, 12)
    print("rio tiler bands: ", img.count)
with Reader(file_path) as cog:
    img = cog.tile(2100, 1369, 12, indexes=[1, 2, 3, 4])
    print("rio tiler forced bands: ", img.count)

with rasterio.open(file_path) as src:
    print("rasterio bands: ", src.count)
``` 
    
returns:
rio tiler bands:  3
rio tiler forced bands:  4
rasterio bands:  4

is there a specific reason for only reading 3 bands by default?

Regards

**Comment by vincentsarago** [comment]
Hi @hambsch 

Because it's a RGB-Alpha tiff, rio-tiler will be smart enough to convert the alpha band into the mask 

you can check the `img.mask`, which should represent the `alpha` band values.

**Reply by hambsch** [reply]
Hi @vincentsarago 

the mask band is not suitable to get alpha transparency:

```
from rio_tiler.io import Reader

file_path = r"C:\RMAbuild\Projects\weiss-geoloket_api\data\route\ant-all-all-50-netto_2010.tif"

with Reader(file_path) as cog:
    img = cog.tile(2100, 1369, 12)
    print("mask value at cell 255,255:", img.mask[255,255])
    print("rio tiler bands: ", img.count)
with Reader(file_path) as cog:
    img = cog.tile(2100, 1369, 12, indexes=[1, 2, 3, 4])
    print("alpha value at cell 255,255:" ,   img.data[3,255, 255])
    print("rio tiler forced bands: ", img.count)
``` 

mask value at cell 255,255: 255
rio tiler bands:  3
alpha value at cell 255,255: 57
rio tiler forced bands:  4

the mask array is only 0 or 255, 0 where de source band is nodata and 255 where it is any other value.

**Reply by vincentsarago** [reply]
so you have an alpha band with partially masked value (`57`), this is not supported in rio-tiler and the value will be either 0 or 255. 

this was changed in https://github.com/cogeotiff/rio-tiler/blob/main/CHANGES.md#500-2023-06-01, more and this topic can be found over https://github.com/rasterio/rasterio/issues/1721#issuecomment-586547617

**Reply by vincentsarago** [reply]
it all comes down to https://github.com/cogeotiff/rio-tiler/blob/83e8fc4ed765445d1aa0267cc3bccee15664b9b3/rio_tiler/reader.py#L216

The problem (as implementer) is to know what `partial` values means for the user. By default `numpy` will convert `57` to `True` while YOU would expect to be `False` 🤷‍♂️ 

We could add a configuration parameter to let the user control this. I'm happy to review PR 🙏

**Reply by hambsch** [reply]
I just expect a 4th band with 57 in it, the same as rasterio.read().
So the only way to read the correct 4 bands as raw data is forcing the indexes with indexes=[1, 2, 3, 4]?
Then i will implement it like this in my code.
Thank you for the clarification.
